File: LEEStarRatingDemo-OC/LEEStarRating/LEEStarRating/LEEStarRating.m

#import "LEEStarRating.h"

#define STARCOUNT starCount

#define SELFX self.frame.origin.x

#define SELFY self.frame.origin.y

#define SELFWIDTH CGRectGetWidth(self.frame)

#define SELFHEIGHT CGRectGetHeight(self.frame)

@interface LEEStarRating ()

@property (nonatomic , strong ) UIView *checkedImagesView; // 已选中星星图片视图

@property (nonatomic , strong ) UIView *uncheckedImagesView; // 未选中星星图片视图

@end

@implementation LEEStarRating
{
    CGSize starSize;
    
    NSUInteger starCount;
}

- (instancetype)initWithFrame:(CGRect)frame
{
    return [self initWithFrame:frame Count:5];
}

- (instancetype)initWithFrame:(CGRect)frame Count:(NSUInteger)count
{
    self = [super initWithFrame:frame];
    if (self) {
        
        starCount = count;
        
        // 初始化数据
        
        [self initData];
        
        // 初始化子视图
        
        [self initSubViews];
        
        // 设置布局
        
        [self configLayout];
    }
    return self;
}

- (void)initData{
    
    // 设置默认值
    
    _spacing = 5.0f;
    
    _minimumScore = 0.0f;
    
    _maximumScore = 5.0f;
    
    _type = RatingTypeWhole;
}

- (void)initSubViews{
    
    // 初始化已选中和未选中图片视图
    
    _uncheckedImagesView = [[UIView alloc] init];
    
    [self addSubview:_uncheckedImagesView];
    
    _checkedImagesView = [[UIView alloc] init];
    
    _checkedImagesView.clipsToBounds = YES;
    
    [self addSubview:_checkedImagesView];
    
    // 循环初始化已选中和未选中子视图
    
    for (NSInteger i = 0; i < STARCOUNT; i++) {
        
        UIImageView *uncheckedImage = [[UIImageView alloc] initWithImage:self.uncheckedImage];
        
        [self.uncheckedImagesView addSubview:uncheckedImage];
        
        UIImageView *checkedImage = [[UIImageView alloc] initWithImage:self.checkedImage];
        
        [self.checkedImagesView addSubview:checkedImage];
    }
    
}

- (void)configLayout{
    
    NSAssert(STARCOUNT * self.spacing < SELFWIDTH, @"间距过长 已超出视图大小");
    
    // 计算星星大小
    
    CGFloat size = (SELFWIDTH - (STARCOUNT + 1) * self.spacing) / STARCOUNT ? : 0;
    
    starSize = CGSizeMake(size, size);
    
    // 如果当前高度小于星星高度 则设置当前高度为星星高度
    
    if (SELFHEIGHT < size) self.frame = CGRectMake(SELFX, SELFY, SELFWIDTH, size);
    
    // 设置已选中和未选中视图
    
    self.uncheckedImagesView.frame = CGRectMake(0, 0, SELFWIDTH, starSize.height);
    
    self.uncheckedImagesView.center = CGPointMake(SELFWIDTH * 0.5f, SELFHEIGHT * 0.5f);
    
    self.checkedImagesView.frame = self.uncheckedImagesView.frame;
    
    for (NSInteger i = 0; i < STARCOUNT; i++) {
        
        UIImageView *uncheckedImage = self.uncheckedImagesView.subviews[i];
        
        UIImageView *checkedImage = self.checkedImagesView.subviews[i];
        
        CGRect imageFrame = CGRectMake(i ? (starSize.width + self.spacing) * i + self.spacing : self.spacing, 0, starSize.width, starSize.height);
        
        uncheckedImage.frame = imageFrame;
        
        checkedImage.frame = imageFrame;
    }
    
}

#pragma mark - Setter

- (void)setSpacing:(CGFloat)spacing{
    
    _spacing = spacing;
    
    [self configLayout];
}

- (void)setScoreEnabled:(BOOL)scoreEnabled{
    
    _scoreEnabled = scoreEnabled;
    
    if (scoreEnabled) {
        
        UITapGestureRecognizer * tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapGestureEvent:)];
        
        UIPanGestureRecognizer * panGesture = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(panGestureEvent:)];
        
        [self addGestureRecognizer:tapGesture];
        
        [self addGestureRecognizer:panGesture];
    }
    
}

- (void)setUncheckedImage:(UIImage *)uncheckedImage{
    
    _uncheckedImage = uncheckedImage;
    
    [self.uncheckedImagesView.subviews enumerateObjectsUsingBlock:^(__kindof UIImageView * _Nonnull imageView, NSUInteger idx, BOOL * _Nonnull stop) {
       
        imageView.image = uncheckedImage;
    }];
}

- (void)setCheckedImage:(UIImage *)checkedImage{
    
    _checkedImage = checkedImage;
    
    [self.checkedImagesView.subviews enumerateObjectsUsingBlock:^(__kindof UIImageView * _Nonnull imageView, NSUInteger idx, BOOL * _Nonnull stop) {
        
        imageView.image = checkedImage;
    }];
}

- (void)setCurrentScore:(CGFloat)currentScore{
    
    NSAssert(self.minimumScore <= currentScore, @"当前分数小于最小分数");
    
    NSAssert(self.maximumScore >= currentScore, @"当前分数大于最大分数");
    
    if (currentScore < self.minimumScore) currentScore = self.minimumScore;
    
    if (currentScore > self.maximumScore) currentScore = self.maximumScore;
    
    [self updateStarView:(currentScore - self.minimumScore) / (self.maximumScore - self.minimumScore)];
}

- (void)setMinimumScore:(CGFloat)minimumScore{
    
    _minimumScore = minimumScore;
    
    NSAssert(minimumScore >= 0, @"最小分数不能小于0");
}

- (void)setMaximumScore:(CGFloat)maximumScore{
    
    _maximumScore = maximumScore;
    
    NSAssert(maximumScore > 0, @"最大分数不能小于0");
    
    NSAssert(maximumScore > self.minimumScore, @"最大分数不能小于最小分数");
}

#pragma mark - Action

- (void)tapGestureEvent:(UITapGestureRecognizer *)tap{
    
    CGPoint point = [tap locationInView:self];
    
    [self updateStarView:[self pointToRatio:point]];
}

- (void)panGestureEvent:(UITapGestureRecognizer *)pan{
    
    CGPoint point = [pan locationInView:self];
    
    [self updateStarView:[self pointToRatio:point]];
}

#pragma mark - Tools

- (CGFloat)pointToRatio:(CGPoint)point{

    // 坐标 转 所选中的比例
    
    CGFloat ratio = 0.0f;
    
    if (self.spacing > point.x) {
        
        ratio = 0.0f;
        
    } else if (SELFWIDTH - self.spacing < point.x) {
        
        ratio = 1.0f;
        
    } else {
        
        /* 比例转换
         *
         * 当前点击位置在当前视图宽度中的比例 转换为 当前点击星星位置在全部星星宽度中的比例
         * 当前点击位置去除其中的间距宽度等于星星的宽度 point.x - 间距 = 所选中的星星宽度
         * 所选中的星星宽度 / 所有星星宽度 = 当前选中的比例
         */
        
        CGFloat itemWidth = self.spacing + starSize.width;
        
        CGFloat icount = point.x / itemWidth;
        
        NSInteger count = floorf(icount);
        
        CGFloat added = (itemWidth * (icount - count));
        
        added = added >= self.spacing ? self.spacing : added;
        
        CGFloat x = point.x - self.spacing * count - added;
        
        ratio = x / (starSize.width * STARCOUNT);
    }

    return ratio;
}

#pragma mark - 更新星星视图 传入当前所选中的比例值

- (void)updateStarView:(CGFloat)ratio{
    
    if (ratio < 0) return;
    
    if (ratio > 1) return;
    
    CGFloat width = 0.0f;
    
    // 根据类型计算比例和宽度
    
    switch (self.type) {
            
        case RatingTypeWhole:
        {
            ratio = ceilf(STARCOUNT * ratio);
            
            width = starSize.width * ratio + (self.spacing * roundf(ratio));
        }
            break;
            
        case RatingTypeHalf:
        {
            
            ratio = STARCOUNT * ratio;
            
            CGFloat s = (ratio) - floorf(ratio);
            
            if (s >= 0.5f) {
                
                ratio = floorf(ratio) + 1.0f;
            }
            
            if (s < 0.5f && s > 0) {
                
                ratio = floorf(ratio) + 0.5f;
            }
            
            width = starSize.width * ratio + (self.spacing * roundf(ratio));
        }
            break;
            
        case RatingTypeUnlimited:
            
            ratio = STARCOUNT * ratio;
            
            width = starSize.width * ratio + (self.spacing * ceilf(ratio));
            
            break;
            
        default:
            break;
    }
    
    // 设置宽度
    
    if (width < 0) width = 0;
    
    if (width > SELFWIDTH) width = SELFWIDTH;
    
    CGRect checkedImagesViewFrame = self.checkedImagesView.frame;
    
    checkedImagesViewFrame.size.width = width;
    
    self.checkedImagesView.frame = checkedImagesViewFrame;
    
    // 设置当前分数
    
    CGFloat currentScore = ratio / STARCOUNT * (self.maximumScore - self.minimumScore) + self.minimumScore;
    
    if (currentScore < self.minimumScore) currentScore = self.minimumScore;
    
    if (currentScore > self.maximumScore) currentScore = self.maximumScore;
    
    if (_currentScore != currentScore) {
        
        _currentScore = currentScore;
        
        if (self.currentScoreChangeBlock) self.currentScoreChangeBlock(self.currentScore);
    }
    
}

@end

File: LEEStarRatingDemo-Swift/LEEStarRating/LEEStarRating/LEEStarRating.swift
import UIKit

public class LEEStarRating: UIView {
    
    public struct Configuration {
        public enum Mode {
            case whole      // 整颗星星
            case half       // 半颗星星
            case unlimited  // 无限制
        }
        
        /// 星星个数
        public var count: Int = 5
        /// 间距 (默认 5.0)
        public var spacing: Double = 5.0
        /// 最大分数 (默认 5.0)
        public var maximumScore: Double = 5.0
        /// 最小分数 (默认 0.0)
        public var minimumScore: Double = 0.0
        /// 启用点击 (默认 false)
        public var touchEnabled: Bool = false
        /// 启用滑动 (默认 false)
        public var slideEnabled: Bool = false
        /// 评分类型 (默认 .whole)
        public var mode: Mode = .whole
    }
    
    /// 配置
    private let configuration: Configuration
    
    /// 已选中星星图片 (保证大小相同)
    public var checkedImage: UIImage? {
        didSet {
            checkedImagesView.subviews.forEach { view in
                (view as? UIImageView)?.image = checkedImage
            }
        }
    }
    /// 未选中星星图片 (保证大小相同)
    public var uncheckedImage: UIImage? {
        didSet {
            uncheckedImagesView.subviews.forEach { view in
                (view as? UIImageView)?.image = uncheckedImage
            }
        }
    }
    
    /// 当前分数 (默认 0.0)
    public private(set) var currentScore: Double = 0.0
    /// 当前分数变更回调
    public var currentScoreChange: ((Double) -> Void)?
    
    private lazy var checkedImagesView = UIView()
    private lazy var uncheckedImagesView = UIView()
    private var star: CGSize
    
    init(frame: CGRect, configuration: Configuration = .init()) {
        self.configuration = configuration
        let size = (frame.size.width - .init(configuration.count + 1) * configuration.spacing) / .init(configuration.count)
        self.star = .init(width: size, height: size)
        super.init(frame: frame)
        
        isUserInteractionEnabled = configuration.touchEnabled || configuration.slideEnabled
        
        if (configuration.touchEnabled) {
            let tap = UITapGestureRecognizer(target: self, action: #selector(tapAction))
            addGestureRecognizer(tap)
        }
        
        if (configuration.slideEnabled) {
            let pan = UIPanGestureRecognizer(target: self, action: #selector(panAction))
            addGestureRecognizer(pan)
        }
        
        uncheckedImagesView.clipsToBounds = true
        checkedImagesView.clipsToBounds = true
        
        addSubview(uncheckedImagesView)
        addSubview(checkedImagesView)
        
        for _ in 0 ..< configuration.count {
            checkedImagesView.addSubview(UIImageView(image: nil))
            uncheckedImagesView.addSubview(UIImageView(image: nil))
        }
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        
        let size = (frame.size.width - .init(configuration.count + 1) * configuration.spacing) / .init(configuration.count)
        star = .init(width: size, height: size)
        
        uncheckedImagesView.frame = .init(x: 0, y: 0, width: frame.width, height: star.height)
        uncheckedImagesView.center = .init(x: frame.width * 0.5, y: frame.height * 0.5)
        
        checkedImagesView.frame = .init(
            x: 0,
            y: uncheckedImagesView.frame.origin.y,
            width: checkedImagesView.frame.width,
            height: uncheckedImagesView.frame.height
        )
        
        for i in 0 ..< configuration.count {
            let checkedImage = checkedImagesView.subviews[i]
            let uncheckedImage = uncheckedImagesView.subviews[i]
            let spacing = configuration.spacing
            let new = CGRect(
                x: i == 0 ? spacing : (star.width + spacing) * .init(i) + spacing,
                y: 0,
                width: star.width,
                height: star.height
            )
            checkedImage.frame = new
            uncheckedImage.frame = new
        }
    }
    
    @objc
    private func tapAction(_ sender: UITapGestureRecognizer) {
        let point = sender.location(in: self)
        update(ratio: ratio(at: point))
    }
    
    @objc
    private func panAction(_ sender: UIPanGestureRecognizer) {
        let point = sender.location(in: self)
        update(ratio: ratio(at: point))
    }
    
    /// 设置当前分数 不能超出最大最小分数范围
    /// - Parameter score: 分数
    public func set(current score: Double) {
        guard score >= configuration.minimumScore, score <= configuration.maximumScore else {
            return
        }
        update(ratio: (score - configuration.minimumScore) / (configuration.maximumScore - configuration.minimumScore))
    }
}

extension LEEStarRating {
    
    private func update(ratio value: Double) {
        guard value >= 0, value <= 1 else {
            return
        }
        
        let ratio: Double
        var width: Double
        
        switch configuration.mode {
        case .whole:
            let temp = (.init(configuration.count) * value).rounded(.up)
            ratio = temp
            width = star.width * temp + (configuration.spacing * temp.rounded())
            
        case .half:
            var temp = (.init(configuration.count) * value)
            let down = temp.rounded(.down)
            let diff = temp - down
            temp = diff > 0.5 ? down + 1.0 : (diff >= 0.001 ? down + 0.5 : down)
            ratio = temp
            width = star.width * temp + (configuration.spacing * temp.rounded())
            
        case .unlimited:
            let temp = (.init(configuration.count) * value)
            ratio = temp
            width = star.width * temp + (configuration.spacing * temp.rounded(.up))
        }
        
        width = max(min(width, frame.width), 0)
        
        checkedImagesView.frame.size.width = width
        
        let score = configuration.maximumScore - configuration.minimumScore
        var current = (ratio / .init(configuration.count)) * score + configuration.minimumScore
        current = max(min(current, configuration.maximumScore), configuration.minimumScore)
        if currentScore != current {
            currentScore = current
            currentScoreChange?(current)
        }
    }
    
    private func ratio(at point: CGPoint) -> Double {
        let spacing = configuration.spacing
        let ratio: Double
        
        if spacing > point.x {
            ratio = 0.0
            
        } else if frame.width - spacing < point.x {
            ratio = 1.0
            
        } else {
            /* 比例转换
             *
             * 当前点击位置在当前视图宽度中的比例 转换为 当前点击星星位置在全部星星宽度中的比例
             * 当前点击位置去除其中的间距宽度等于星星的宽度 point.x - 间距 = 所选中的星星宽度
             * 所选中的星星宽度 / 所有星星宽度 = 当前选中的比例
             */
            
            let single = spacing + star.width
            let temp = point.x / single
            let count = temp.rounded(.down)
            var added = single * (temp - count)
            added = added >= spacing ? spacing : added
            let x = point.x - spacing * count - added
            ratio = x / (star.width * .init(configuration.count))
        }
        
        // 解决浮点精度丢失问题
        guard let decimal = Decimal(string: .init(format: "%0.4lf", ratio)) else { return ratio }
        return NSDecimalNumber(decimal: decimal).doubleValue
    }
}
